cmd/zfs_object_agent/src/block_allocator.rs

```diff
@@ -87,8 +146,604 @@ impl BlockAllocator {
+            Some(id) => {
+                let slab_extent = self.slabs[id.0 as usize].allocate(request_size);
+                match slab_extent {
+                    Some(extent) => {
+                        self.dirty_slab_id(id);
+                        let slab_internal_offset = extent.location.offset;
+                        Some(Extent {
+                            location: DiskLocation {
+                                offset: self.slab_offset_from_slab_id(id) + slab_internal_offset,
+                            },
+                            size: request_size as usize,
+                        })
+                    }
+                    None => {
+                        //
+                        // XXX: At this point we should look for the next suitable slab or
+                        //      curve a new one as needed.
+                        //
+                        sorted_slabs.advance();
+                        None
+                    }
+                }
+            }
+            None => self.allocate_from_new_slab(request_size),
+        }
+    }
+
+    pub fn free(&mut self, extent: Extent) {
+        let slab_id = self.slab_id_from_extent(extent);
+        let slab_offset = self.slab_offset_from_slab_id(slab_id);
+        let slab_internal_offset = extent.location.offset - slab_offset;
+        self.slabs[slab_id.0 as usize].free(Extent {
+            location: DiskLocation {
+                offset: slab_internal_offset,
+            },
+            size: extent.size,
+        });
+        self.dirty_slab_id(slab_id);
+    }
+
+    pub async fn flush(&mut self) -> BlockAllocatorPhys {
+        let mut flush_extend_slabs = false;
+        let mut bitmap_blocksizes_to_flush = HashSet::new();
+
+        for slab_id in self.dirty_slabs.iter() {
+            let slab_offset = self.slab_offset_from_slab_id(*slab_id);
+            let slab = &mut self.slabs[slab_id.0 as usize];
+            slab.flush_on_spacemap(&mut self.spacemap, slab_offset);
+            slab.undirty();
+
+            match &slab.info {
+                SlabType::BitmapBased(slab_info) => {
+                    bitmap_blocksizes_to_flush.insert(slab_info.block_size);
+                }
+                SlabType::ExtentBased(_) => flush_extend_slabs = true,
+                SlabType::Free => panic!("free slab was dirtied"),
+            }
+
+            if slab.free_space() == (self.slab_size as u64) {
+                self.free_slabs.push(slab.id);
+                *slab = Slab::new_empty(*slab_id);
+            }
+        }
+        self.dirty_slabs.clear();
+
+        if flush_extend_slabs {
+            self.extent_slabs.reset(&self.slabs);
+        }
+        for block_size in bitmap_blocksizes_to_flush.iter() {
+            let bucket: u32 = block_size / self.bitmap_bucket_size;
+            self.bitmap_slabs
+                .get_mut(&bucket)
+                .unwrap()
+                .reset(&mut self.slabs);
```

Comment:
```suggestion
                .reset(&self.slabs);
```
I think the compiler should warn you about using `&mut` where `&` will suffice